<a href="https://colab.research.google.com/github/nandadwilestari/Data-Mining/blob/main/UTS_Data_Mining__Nanda_Dwi_Lestari.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Import Library

In [33]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

from sklearn.ensemble import ExtraTreesClassifier

## Penjelasan:

Digunakan untuk mengimpor library yang diperlukan dalam proses analisis data dan machine learning.


# 2. Load Dataset

In [3]:
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')

print(train.head())
print(test.head())

   fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.3             0.740         0.08             1.7      0.094   
1            8.1             0.575         0.22             2.1      0.077   
2           10.1             0.430         0.40             2.6      0.092   
3           12.9             0.500         0.55             2.8      0.072   
4            8.4             0.360         0.32             2.2      0.081   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 10.0                  45.0  0.99576  3.24       0.50   
1                 12.0                  65.0  0.99670  3.29       0.51   
2                 13.0                  52.0  0.99834  3.22       0.64   
3                  7.0                  24.0  1.00012  3.09       0.68   
4                 32.0                  79.0  0.99640  3.30       0.72   

   alcohol  quality    Id  
0      9.8        5  1366  
1      9.2        5   103  
2 

## Penjelasan:

Digunakan untuk membaca dataset training dan testing menggunakan pandas. Dataset berhasil dimuat dan dapat dilihat struktur awal datanya.

# 3. Mengecek Informasi Dataset

In [4]:
print(train.info())
print(test.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 286 entries, 0 to 285
D

## Penjelasan:

Digunakan untuk melihat:

*   jumlah data
*   tipe data
*  jumlah non-null
*   struktur dataset

## Hasil:

Dataset training memiliki:
*   857 baris
*   13 kolom

Dataset testing memiliki:
*   286 baris
*   12 kolom

Seluruh fitur bertipe numerik (float64 dan int64) serta tidak terdapat masalah tipe data sehingga dataset siap digunakan untuk proses machine learning.

# 4. Mengecek Missing Value

In [5]:
print(train.isnull().sum())
print(test.isnull().sum())

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
Id                      0
dtype: int64
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
Id                      0
dtype: int64


## Penjelasan:

Digunakan untuk mengecek apakah terdapat missing value pada dataset.

## Hasil:

Tidak terdapat missing value pada dataset training maupun testing karena seluruh kolom memiliki nilai 0. Artinya data sudah bersih dan dapat langsung digunakan tanpa proses imputasi tambahan.

# 5. Memisahkan Feature dan Target

In [44]:
# Feature
X = train.drop(columns=['quality', 'Id'])

# Target
y = train['quality']

# Data testing
X_test = test.drop(columns=['Id'])

## Penjelasan:

Digunakan untuk memisahkan variabel input (feature) dan target.

*   Variabel X berisi seluruh fitur kimia wine.
*   Variabel y berisi label kualitas wine (quality).
*   Variabel X_test digunakan sebagai data testing untuk proses prediksi.

Data berhasil dipisahkan sehingga siap digunakan untuk training model.

# 6. Split Data Training dan Validation

In [45]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

## Penjelasan:

Digunakan untuk membagi data menjadi data training dan validation. 80% data digunakan untuk training dan 20% data digunakan untuk validation. Pembagian data dilakukan menggunakan stratify=y sehingga distribusi kelas kualitas wine tetap seimbang pada data training dan validation.


# 7. Membuat Pipeline Model

In [37]:
model = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler()),
    ('classifier', ExtraTreesClassifier(
        n_estimators=500,
        random_state=42
    ))
])

## Penjelasan:

Pipeline digunakan untuk menggabungkan preprocessing dan model dalam satu proses otomatis. Tahapan pipeline adalah sebagai berikut:

Imputer → menangani missing value
StandardScaler → normalisasi data
ExtraTreesClassifier → model klasifikasi

## Hasil:

Pipeline berhasil dibuat dengan tahapan:
1.   SimpleImputer
2.   StandardScaler
3.   ExtraTreesClassifier

Model machine learning siap digunakan untuk proses training dan evaluasi.


# 8. Training Model

In [46]:
model.fit(X_train, y_train)

Pipeline(steps=[('imputer', SimpleImputer(strategy='median')),
                ('scaler', StandardScaler()),
                ('classifier',
                 ExtraTreesClassifier(n_estimators=500, random_state=42))])

## Penjelasan:

Digunakan untuk melatih model menggunakan data training.

## Hasil:

Model berhasil dilatih menggunakan data training. Model berhasil mempelajari pola hubungan antara fitur kimia wine dan kualitas wine.

# 9. Evaluasi Model

In [47]:
# Prediksi validation
pred_valid = model.predict(X_valid)

# Accuracy
acc = accuracy_score(y_valid, pred_valid)
print('Accuracy :', acc)

# Confusion Matrix
print(confusion_matrix(y_valid, pred_valid))

# Classification Report
print(classification_report(y_valid, pred_valid))

Accuracy : 0.6104651162790697
[[ 0  0  0  1  0  0]
 [ 0  0  5  0  0  0]
 [ 0  0 58 14  1  0]
 [ 0  0 26 39  3  0]
 [ 0  0  1 14  7  0]
 [ 0  0  0  0  2  1]]
              precision    recall  f1-score   support

           3       0.00      0.00      0.00         1
           4       0.00      0.00      0.00         5
           5       0.64      0.79      0.71        73
           6       0.57      0.57      0.57        68
           7       0.54      0.32      0.40        22
           8       1.00      0.33      0.50         3

    accuracy                           0.61       172
   macro avg       0.46      0.34      0.36       172
weighted avg       0.59      0.61      0.59       172



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


## Penjelasan:

Digunakan untuk mengukur performa model menggunakan data validation.

## Hasil:

Model memperoleh:
*   Accuracy sebesar sekitar 61%
*   Precision terbaik pada kelas quality 5 dan 6
*   Recall tertinggi terdapat pada quality 5

Confusion matrix menunjukkan bahwa sebagian besar prediksi berada pada kelas kualitas 5 dan 6 karena jumlah data pada kelas tersebut lebih dominan. Classification report menunjukkan performa model cukup baik dalam memprediksi kualitas wine meskipun beberapa kelas minoritas masih sulit diprediksi.


# 10. Cross Validation

In [40]:
cv_score = cross_val_score(
    model,
    X,
    y,
    cv=5,
    scoring='accuracy'
)

print('Cross Validation Accuracy :', cv_score.mean())

Cross Validation Accuracy : 0.6475724194206445


## Penjelasan:

Digunakan untuk mengecek kestabilan model menggunakan validasi silang.

## Hasil:

Model memperoleh nilai cross validation accuracy sebesar sekitar 64.76%. Hal ini menunjukkan bahwa model memiliki performa yang cukup stabil dan mampu melakukan generalisasi dengan baik pada data baru.


# 11. Training Model Menggunakan Seluruh Data

In [41]:
model.fit(X, y)

Pipeline(steps=[('imputer', SimpleImputer(strategy='median')),
                ('scaler', StandardScaler()),
                ('classifier',
                 ExtraTreesClassifier(n_estimators=500, random_state=42))])

## Penjelasan:

Model dilatih kembali menggunakan seluruh data training. Dengan menggunakan seluruh data, model memperoleh informasi lebih lengkap sehingga performa prediksi pada data testing menjadi lebih optimal.


# 12. Prediksi Data Testing

In [48]:
pred_test = model.predict(X_test)

print(pred_test)

[5 6 5 5 5 6 5 5 6 5 7 6 5 6 5 5 6 5 5 7 6 6 5 7 5 5 6 6 5 6 5 5 7 6 5 5 6
 6 6 5 7 6 6 5 5 5 5 6 5 6 7 5 6 7 6 6 6 6 6 5 5 6 6 7 6 6 6 5 5 6 6 5 5 6
 5 5 5 6 5 6 5 6 6 5 5 6 5 5 6 6 5 6 5 5 5 5 6 6 5 5 6 7 5 6 5 5 6 6 5 7 6
 5 5 6 5 5 6 6 6 5 5 5 5 6 5 5 6 6 5 6 5 5 7 6 6 5 5 5 6 6 7 5 5 5 6 5 5 5
 5 5 6 5 6 5 6 5 5 5 7 6 5 5 6 6 6 5 5 6 6 5 6 5 6 6 5 5 7 6 6 5 6 5 6 5 5
 5 6 5 6 6 6 6 6 6 6 6 5 6 7 6 5 6 5 5 6 7 5 5 6 6 5 6 5 6 5 6 6 5 7 6 6 6
 5 6 5 6 6 5 6 6 6 7 5 5 6 6 5 6 5 5 7 5 6 5 5 5 5 5 6 5 5 5 6 6 5 6 6 5 5
 5 5 6 7 6 5 5 6 6 5 6 6 7 5 6 5 6 5 6 5 5 5 6 5 5 6 7]


## Penjelasan:

Digunakan untuk memprediksi kualitas wine pada data testing.

## Hasil:

Model berhasil menghasilkan prediksi quality untuk setiap data wine pada dataset testing.
Hasil prediksi didominasi oleh nilai quality 5 dan 6 karena kedua kelas tersebut merupakan kelas yang paling banyak muncul pada dataset training.

# 13. Membuat File Submission CSV

In [30]:
submission = pd.DataFrame({
    'Id': test['Id'],
    'Quality': pred_test
})

print(submission.to_string(index=False))


  Id  Quality
 222        5
1514        6
 417        5
 754        5
 516        5
1120        6
 180        5
  82        5
 632        6
 592        5
 501        7
1586        6
1096        6
1101        6
1488        5
 700        5
 631        6
1022        5
1420        5
 882        7
1170        6
 911        6
 124        5
1417        7
 681        5
  63        5
1010        7
 418        6
 399        5
1265        6
 999        5
 452        6
 347        7
 778        6
 607        5
  32        5
 511        6
1035        6
1270        6
 723        5
 838        7
 485        6
 142        6
1470        5
1238        5
 158        5
 426        6
 344        6
 637        5
1578        6
 944        7
 469        5
 148        6
 279        7
1144        6
 509        6
1073        6
1423        6
 280        6
 848        5
 650        5
1092        6
  76        6
1302        7
1148        6
 100        6
 533        6
1328        5
 253        5
1344        6
 544  

## Penjelasan:

Digunakan untuk membuat format file submission

## Hasil:

Dataframe submission berhasil dibuat dengan 2 kolom:
*   Id
*   Quality


# 14. Menyimpan Hasil Prediksi

In [49]:
submission.to_csv(
    'hasilprediksi_3digitNIMterakhir.csv',
    index=False,
    sep=','
)

## Penjelasan:

Digunakan untuk menyimpan hasil prediksi ke dalam file CSV.


# 15. Mendownload Hasil Prediksi

In [50]:
from google.colab import files

files.download('hasilprediksi_3digitNIMterakhir.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# Penjelasan:
files.download() digunakan untuk mengunduh file dari Colab ke laptop.


# Kesimpulan


Dataset wine quality berhasil diproses tanpa error.
Missing value ditangani menggunakan imputasi median.
Feature scaling dilakukan menggunakan StandardScaler.
Model terbaik yang digunakan adalah ExtraTreesClassifier.
Model menghasilkan akurasi sekitar 64%–65%.
Hasil prediksi berhasil disimpan ke dalam file CSV sesuai format pengumpulan.